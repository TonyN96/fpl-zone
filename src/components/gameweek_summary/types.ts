export interface SummaryStatType {
  label: string;
  teamCode?: number;
  playerName?: string;
  value?: string | number;
}
