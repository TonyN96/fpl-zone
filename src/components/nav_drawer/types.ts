export interface MenuItemType {
  id: string;
  label: string;
  type: "nav" | "auth";
  onClick: () => void;
}
