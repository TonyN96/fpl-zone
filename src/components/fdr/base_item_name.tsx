import React from "react";
import { Box, Typography } from "@mui/material";
import { getTeamCrestImageUrl } from "helpers";
import { has } from "lodash";
import { Player, Team } from "types";

import { BaseItem } from "./fdr";

interface BaseItemNameProps {
  baseItem: BaseItem;
}

export const BaseItemName = ({ baseItem }: BaseItemNameProps): JSX.Element => {
  const isPlayerTable = has(baseItem, "web_name");

  const name = isPlayerTable ? (baseItem as Player).web_name : (baseItem as Team).name;
  const teamCode = isPlayerTable ? (baseItem as Player).team_code : (baseItem as Team).code;

  return (
    <Box
      alignItems='center'
      data-testid={`base-item-${baseItem.id}`}
      display='flex'
      gap={1}
      paddingLeft={1}
      whiteSpace='nowrap'
    >
      <img
        alt={`${teamCode}-crest`}
        data-testid={`base-item-crest-img-${baseItem.id}`}
        height='22px'
        src={getTeamCrestImageUrl(teamCode)}
      />
      <Box overflow='hidden'>
        <Typography
          className='text-ellipsis'
          data-testid={`base-item-text-${baseItem.id}`}
          fontWeight={500}
          variant='body1'
        >
          {name.toUpperCase()}
        </Typography>
      </Box>
    </Box>
  );
};
