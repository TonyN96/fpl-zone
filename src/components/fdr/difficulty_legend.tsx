import React from "react";
import { Box, Typography } from "@mui/material";

export const fdrColours = {
  1: "#09BA59",
  2: "#93E02D",
  3: "#F5CF38",
  4: "#DE7628",
  5: "#FF193C",
};

export default function DifficultyLegend(): JSX.Element {
  return (
    <Box display="flex" sx={{ alignItems: "center", minHeight: "50px" }}>
      <Typography sx={{ mr: 1 }}>Difficulty legend:</Typography>
      {Object.keys(fdrColours).map((colour) => {
        return (
          <Box
            component="div"
            key={colour}
            sx={{ backgroundColor: fdrColours[colour], p: "0 10px" }}
          >
            {colour}
          </Box>
        );
      })}
    </Box>
  );
}
