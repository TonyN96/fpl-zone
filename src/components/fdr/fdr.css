.MuiTableCell-root {
  padding: 0;
}
