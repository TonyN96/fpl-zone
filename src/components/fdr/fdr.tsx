import React, { useContext, useEffect, useState } from "react";
import {
  Box,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";
import { getGameweekFixtures } from "api/fpl_api_provider";
import { AppDataContext } from "app_content";
import _ from "lodash";
import { AppData, Fixture, Gameweek, Player, Team } from "types";

import { LoadingMessage } from "components/layout";

import DifficultyLegend from "./difficulty_legend";
import FixtureBox from "./fixture_box";

export type BaseItem = Player | Team;

interface FdrTableProps {
  players?: Player[];
}

export default function FdrTable({ players }: FdrTableProps): JSX.Element {
  const [nextFiveGameweekFixtures, setNextFiveFixtures] = useState<Fixture[][]>([]);
  const appData = useContext(AppDataContext) as AppData;
  const teams = appData.teams;
  const baseItem = players || teams;
  const nameColumnTitle = players ? "Player" : "Team";
  const allGameweeks = appData.events;
  const nextGameweek = allGameweeks.find((gw) => gw.is_next) as Gameweek;
  const nextFiveGameweeks: number[] = [];
  // eslint-disable-next-line no-loops/no-loops
  for (let x = nextGameweek.id; x <= 38 && nextFiveGameweeks.length < 5; x++) {
    nextFiveGameweeks.push(x);
  }

  useEffect(() => {
    const fetchNextFiveGameweekFixtures = async (): Promise<void> => {
      const nextFiveGameweekFixtures: Fixture[][] = await Promise.all(
        nextFiveGameweeks.map(async (gameweek) => await getGameweekFixtures(gameweek))
      );

      setNextFiveFixtures(nextFiveGameweekFixtures);
    };

    fetchNextFiveGameweekFixtures();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const renderBaseItemName = (baseItem: BaseItem): JSX.Element => {
    const name = players ? (baseItem as Player).web_name : (baseItem as Team).name;
    const teamId = players ? (baseItem as Player).team_code : (baseItem as Team).code;
    const testId = `base-item-${name}`;
    return (
      <Box
        sx={{
          display: "flex",
          alignItems: "center",
          columnGap: 1,
          ml: 0.5,
          whiteSpace: "nowrap",
        }}
        data-testid={testId}
      >
        <img
          src={`${process.env.PUBLIC_URL}/assets/images/crests/${teamId}.png`}
          alt="crest-img"
          height="22px"
        />
        <Box
          sx={{
            display: "block",
            whiteSpace: "nowrap",
            overflow: "hidden",
            textOverflow: "ellipsis",
          }}
        >
          <Typography sx={{ overflow: "hidden", textOverflow: "ellipsis", whiteSpace: "nowrap" }}>
            {name}
          </Typography>
        </Box>
      </Box>
    );
  };

  const getTeamById = (teamId: number): string | undefined => {
    const team = teams.find((t) => t.id === teamId);
    return team?.short_name;
  };

  const getNextFiveTeamFixtures = (baseItem: BaseItem, fixtures: Fixture[][]): Fixture[][] => {
    const teamId = players ? (baseItem as Player).team : (baseItem as Team).id;
    return fixtures.map((gameweek) => {
      return gameweek.filter((f) => f.team_h === teamId || f.team_a === teamId);
    });
  };

  const renderRow = (baseItem: BaseItem, index: number): JSX.Element => {
    const teamFixtures = getNextFiveTeamFixtures(baseItem, nextFiveGameweekFixtures);
    const testId = `fixture-row-${baseItem.id}`;
    return (
      <TableRow
        key={index}
        sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
        data-testid={testId}
      >
        <TableCell component="th" scope="row" key={index}>
          {renderBaseItemName(baseItem)}
        </TableCell>
        {_.map(teamFixtures, (fixtures, key) => (
          <FixtureBox
            fixtures={fixtures}
            baseItem={baseItem}
            isPlayerTable={!!players}
            key={key}
            getTeamById={getTeamById}
          />
        ))}
      </TableRow>
    );
  };

  return _.isEmpty(nextFiveGameweekFixtures) ? (
    <Box sx={{ display: "flex", justifyContent: "center", alignItems: "center", height: "100%" }}>
      <LoadingMessage message="Fetching fixture data.." />
    </Box>
  ) : (
    <Box
      display="flex"
      flexDirection="column"
      justifyContent="center"
      alignItems="center"
      overflow="hidden"
      height="100%"
      sx={{ "& .MuiTableContainer-root": { height: "100%" } }}
      data-testid="fdr-container"
    >
      <DifficultyLegend />
      <TableContainer>
        <Table
          aria-label="fdr table"
          sx={{
            tableLayout: "fixed",
            height: "100%",
            flexGrow: "1",
            "& .MuiTableCell-root": { padding: "2px 4px" },
          }}
        >
          <TableHead>
            <TableRow data-testid="table-head-column-title">
              <TableCell sx={{ textAlign: "center" }}>{nameColumnTitle}</TableCell>
              {nextFiveGameweeks.map((gameweekNumber, index) => (
                <TableCell sx={{ textAlign: "center" }} key={index}>
                  GW {gameweekNumber}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {baseItem.map((item: BaseItem, key: number) => renderRow(item, key))}
          </TableBody>
        </Table>
      </TableContainer>
    </Box>
  );
}
