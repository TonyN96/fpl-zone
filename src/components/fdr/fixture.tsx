import React, { useContext } from "react";
import { Box, Tooltip, Typography } from "@mui/material";
import { AppDataContext } from "app_content";
import { getTeamById } from "helpers";
import { AppData, Fixture as FixtureType, Player, Team } from "types";

import { BaseItem, FDR_COLOURS } from "./fdr";

interface FixtureProps {
  fixtures: FixtureType[];
  baseItem: BaseItem;
  isPlayerTable: boolean;
}

interface SingleFixtureProps {
  fixture: FixtureType;
}

export const Fixture = ({
  fixtures,
  baseItem,
  isPlayerTable
}: FixtureProps): JSX.Element => {
  const SingleFixture = ({ fixture }: SingleFixtureProps): JSX.Element => {
    const { teams } = useContext(AppDataContext) as AppData;

    const teamId = isPlayerTable ? (baseItem as Player).team : (baseItem as Team).id;
    const isHome = fixture.team_h === teamId;
    const oppositionId = isHome ? fixture.team_a : fixture.team_h;
    const difficulty = isHome ? fixture.team_h_difficulty : fixture.team_a_difficulty;
    const text = `${getTeamById(oppositionId, teams).short_name} (${isHome ? "H" : "A"})`;

    return (
      <Tooltip
        enterDelay={300}
        placement='top'
        title={text}
      >
        <Box
          className='flex-center'
          data-testid={`fixture-container-bg-${fixture.id}`}
          flexBasis={1}
          flexGrow={1}
          height='100%'
          overflow='hidden'
          sx={{ backgroundColor: `${FDR_COLOURS[difficulty]}` }}
        >
          <Box overflow='hidden' padding={0.5}>
            <Typography className='text-ellipsis'>{text}</Typography>
          </Box>
        </Box>
      </Tooltip>
    );
  };

  return (
    <Box
      alignItems='center'
      data-testid='fixture-container'
      display='flex'
      height='100%'
      justifyContent='space-evenly'
    >
      {fixtures.map((fixture, key) => <SingleFixture fixture={fixture} key={key} />)}
    </Box>
  );
};
