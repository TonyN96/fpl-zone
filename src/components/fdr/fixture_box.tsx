import React from "react";
import { Box, TableCell, Tooltip, Typography } from "@mui/material";
import { Fixture, Player, Team } from "types";

import { fdrColours } from "./difficulty_legend";
import { BaseItem } from "./fdr";

interface FixtureBoxProps {
  fixtures: Fixture[];
  baseItem: BaseItem;
  isPlayerTable: boolean;
  getTeamById: (teamId: number) => string | undefined;
}

export default function FixtureBox({
  fixtures,
  baseItem,
  isPlayerTable,
  getTeamById
}: FixtureBoxProps): JSX.Element {
  const renderBox = (fixture: Fixture, key: number): JSX.Element => {
    const teamId = isPlayerTable ? (baseItem as Player).team : (baseItem as Team).id;
    const isHome = fixture.team_h === teamId;
    const oppositionId = isHome ? fixture.team_a : fixture.team_h;
    const difficulty = isHome ? fixture.team_h_difficulty : fixture.team_a_difficulty;
    const testId = `fix-box-bg-${fixture.id}`;
    const text = `${getTeamById(oppositionId)} (${isHome ? "H" : "A"})`;

    return (
      <Tooltip title={text} placement='top' enterDelay={500} key={key} arrow>
        <Box
          sx={{
            p: 0.5,
            display: "flex",
            backgroundColor: `${fdrColours[difficulty]}`,
            height: "100%",
            flexGrow: 1,
            overflow: "hidden",
            justifyContent: "center",
            alignItems: "center",
            flexBasis: 1
          }}
          key={key}
          data-testid={testId}
        >
          <Box
            sx={{
              p: 0.2,
              textOverflow: "ellipsis",
              whiteSpace: "nowrap",
              overflow: "hidden"
            }}
          >
            <Typography sx={{ textOverflow: "ellipsis", overflow: "hidden" }}>
              {text}
            </Typography>
          </Box>
        </Box>
      </Tooltip>
    );
  };

  return (
    <TableCell scope='row'>
      <Box
        sx={{
          display: "flex",
          height: "100%",
          alignItems: "center",
          justifyContent: "space-evenly"
        }}
        data-testid='fixture-box-container'
      >
        {fixtures.map((fixture, key) => renderBox(fixture, key))}
      </Box>
    </TableCell>
  );
}
