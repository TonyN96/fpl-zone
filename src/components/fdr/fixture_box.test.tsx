import React from "react";
import { render, screen } from "@testing-library/react";
import { Fixture, Player, Team } from "types";

import { fdrColours } from "components/fdr/difficulty_legend";
import { BaseItem } from "components/fdr/fdr";
import FixtureBox from "components/fdr/fixture_box";

import "@testing-library/jest-dom/extend-expect";

import { mockFixtures, mockPlayers, mockTeams } from "../../test/test_data";

describe("Fixture Box Tests", () => {
  let fixtures: Fixture[];
  let baseItem: BaseItem;
  let isPlayerTable: boolean;
  let allTeams: Team[];

  const getTeamById = (teamId: number): string | undefined => {
    const team = allTeams.find((t) => t.id === teamId);
    return team?.short_name;
  };

  const createComponent = (): JSX.Element => {
    return (
      <FixtureBox
        fixtures={fixtures}
        isPlayerTable={isPlayerTable}
        getTeamById={getTeamById}
        baseItem={baseItem}
      />
    );
  };

  beforeEach(() => {
    fixtures = mockFixtures.slice(0, 2);
    allTeams = mockTeams;
  });

  describe("Player fixture box", () => {
    beforeEach(() => {
      [baseItem] = mockPlayers;
      isPlayerTable = true;
    });

    it("Snapshot test", () => {
      const { asFragment } = render(createComponent());
      expect(asFragment()).toMatchSnapshot();
    });

    it("Opposition team abbreviations displayed as expected", () => {
      render(createComponent());

      const fixtureBox = screen.getByTestId("fixture-box-container");

      const teamId = (baseItem as Player).team;

      fixtures.forEach((fix) => {
        const isHome = fix.team_h === teamId;
        const oppositionId = isHome ? fix.team_a : fix.team_h;
        const difficulty = isHome ? fix.team_h_difficulty : fix.team_a_difficulty;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const teamAbbr = getTeamById(oppositionId)!;
        expect(fixtureBox).toHaveTextContent(teamAbbr);

        const testId = `fix-box-bg-${fix.id}`;
        const difficultyColour = fdrColours[difficulty];
        const singleFixtureBox = screen.getByTestId(testId);
        expect(singleFixtureBox).toHaveStyle(`background-color: ${difficultyColour}`);
      });
    });
  });

  describe("Team fixture box", () => {
    beforeEach(() => {
      [baseItem] = mockTeams;
      isPlayerTable = false;
    });

    it("Snapshot test", () => {
      const { asFragment } = render(createComponent());
      expect(asFragment()).toMatchSnapshot();
    });

    it("Opposition team abbreviations displayed as expected", () => {
      render(createComponent());

      const fixtureBox = screen.getByTestId("fixture-box-container");

      const teamId = (baseItem as Team).id;

      fixtures.forEach((fix) => {
        const isHome = fix.team_h === teamId;
        const oppositionId = isHome ? fix.team_a : fix.team_h;
        const difficulty = isHome ? fix.team_h_difficulty : fix.team_a_difficulty;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const teamAbbr = getTeamById(oppositionId)!;
        expect(fixtureBox).toHaveTextContent(teamAbbr);

        const testId = `fix-box-bg-${fix.id}`;
        const difficultyColour = fdrColours[difficulty];
        const singleFixtureBox = screen.getByTestId(testId);
        expect(singleFixtureBox).toHaveStyle(`background-color: ${difficultyColour}`);
      });
    });
  });
});
