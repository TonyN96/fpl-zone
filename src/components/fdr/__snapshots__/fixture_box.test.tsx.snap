// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fixture Box Tests Player fixture box Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1g0fcwx"
    data-testid="fixture-box-container"
  >
    <div
      aria-label="BRE (A)"
      class="MuiBox-root css-32nkmy"
      data-mui-internal-clone-element="true"
      data-testid="fix-box-bg-1"
    >
      <div
        class="MuiBox-root css-19jcvzm"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1k2fol-MuiTypography-root"
        >
          BRE (A)
        </p>
      </div>
    </div>
    <div
      aria-label="LIV (A)"
      class="MuiBox-root css-18be0zc"
      data-mui-internal-clone-element="true"
      data-testid="fix-box-bg-15"
    >
      <div
        class="MuiBox-root css-19jcvzm"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1k2fol-MuiTypography-root"
        >
          LIV (A)
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Fixture Box Tests Team fixture box Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1g0fcwx"
    data-testid="fixture-box-container"
  >
    <div
      aria-label="BRE (A)"
      class="MuiBox-root css-32nkmy"
      data-mui-internal-clone-element="true"
      data-testid="fix-box-bg-1"
    >
      <div
        class="MuiBox-root css-19jcvzm"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1k2fol-MuiTypography-root"
        >
          BRE (A)
        </p>
      </div>
    </div>
    <div
      aria-label="LIV (A)"
      class="MuiBox-root css-18be0zc"
      data-mui-internal-clone-element="true"
      data-testid="fix-box-bg-15"
    >
      <div
        class="MuiBox-root css-19jcvzm"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1k2fol-MuiTypography-root"
        >
          LIV (A)
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
